### DISCUSSION TITLE
[Docs] Firewall ACL and DNS records

### DISCUSSION THREAD

**Discussion by baylanger** [llamafilm/tesla-http-proxy-addon#13] [General]
Hi,

I was reading the docs and was going to make changes to make a PR, then I figured I should ask.

I wanted to confirm the following, note L7 filtering is via Cloudflare:

Create a DNS record tesla.example.com that points to your public IP**
Firewall L4 ACL : allow traffic from any public IP*** to Home Assistant NGINX add-on port 443
Firewall L7 ACL : allow traffic to tesla.example.com/callback
** [edit 3] To prevent rogue actors to hit the add-on, looks like this entry could be tesla-random-name.example.com.
*** this can probably be limited by country based on the location of Tesla servers and possibly limited even more with Tesla Fleet IPs resolved from a list of FQDNs or ASN 394161 filtering if Tesla API IPs are part of their ASN netblocks.

I don't necessarily plan to update the docs with all mentions of the above.

DNS IP entries...  I understand that tesla.example.com home.example.com must be the public IP to reach Home Assistant but is it a requirement for home.example.com to also have a public IP address? For some people, including myself, at home or away I always reach my Home Assistant via a private IP using a VNP (WireGuard) that opens a direct connection on my router. Companion app never reaches Home Assistant over a public IP. Will such a setup break how the proxy/Tesla Fleet API or home.example.com on public Internet DNS must resolve also to a public IP?

[edit # 2] Per the doc I had improperly set the callback origin to home.example.com instead of tesla.example.com , I removed my 1st edit above after realizing I had an error on my setup.

Based on replies, I'll try to make changes to the doc to clarify anything.

Thank you!

**Comment by baylanger** [comment]
> Open this add-on Web UI in the iOS Home Assistant app and click Generate OAuth Token

Can it be changed to something like ...

> Using Home Assistant Companion app or a Web browser, open this add-on and click Generate OAuth Token

**Comment by baylanger** [comment]
> > Open this add-on Web UI in the iOS Home Assistant app and click Generate OAuth Token
> 
> Can it be changed to something like ...
> 
> > Using Home Assistant Companion app or a Web browser, open this add-on and click Generate OAuth Token
>

I just noticed the following:

> Return to the add-on Web UI and click Enroll public key in your vehicle. This will launch the Tesla app where it prompts for approval.

The first step above could be done using a Web browser but not the 2nd one.

**Comment by llamafilm** [comment]
This will not work on a PC because enrolling the public key links you to the Tesla app. I guess it should work on Android too but I don’t have one to test.

You’re right, the home.example.com does not need to be public.

**Reply by baylanger** [reply]
Should the settings form for the add-on say something like:

> Callback Fully Qualified Domain Name, lowercase w/o https, e.g. tesla.example.com"

Obviously, with tesla.example.com that's probably enough and shouldn't confuse anyone 🤞

**Comment by baylanger** [comment]
I was looking into the nginx_tesla.conf and the 2 files in the templates directory.

If someone sets "Allowed Origin tesla.example.com" , is this FQDN defined everywhere in this add-on OR per what I see in nginx config, is it true to say only ' /callback " is relevant for the add-on to work properly? In other words, if home.example.com already points to Home Assistant public IP, is it possible to set "Allowed Origin home.assistant.com" instead of having to create a new "tesla" DNS entry. If I'm offtrack, let me know where to look.

Obviously, like in my case, if someone has no public IP pointing to their Home Assistant, then creating a DNS entry w/ public IP is required.

[edit] rewording...

**Reply by llamafilm** [reply]
The callback _must_ be on port 443.  Some people run their Home Assistant on that port, so that's why I recommend a separate DNS name.  Of course there are many ways to adapt this, but I was trying to think of the simplest approach that would work for most users.

**Comment by baylanger** [comment]
I think I'm even more confused after looking at the code... there must be something I don't quite understand unless there's a typo.

Is DOMAIN in run.py is taken from the following add-on configuration entry:
  "Fully Qualified Domain Name, lowercase w/o https, e.g. home.example.com"

Is the home.example.com the proper example name or should it be tesla.example.com ?

Per [this](https://github.com/baylanger/tesla-http-proxy-addon/blob/d461be2d69596d2d49589ded886260576ad0cacd/tesla_http_proxy/app/run.py#L101C4-L101C11) in run.py:
```
'redirect_uri': f"https://{DOMAIN}/callback"
```

Isn't the callback going to tesla.example.com/callback and not home.example.com ?

**Reply by llamafilm** [reply]
Thanks, I will change the example to tesla.example.com

**Reply by baylanger** [reply]
Fantastic, I thought I was offtrack and this makes me a lot less confused now ;)

Good idea to transfer to discussion, I should have done that first but not all repos have Discussion enable, I didn't think about it.

**Comment by llamafilm** [comment]
To your first question about restricting inbound 443, I don't want to maintain an ever-changing list of undocumented IP addresses for Tesla's servers.  You could try asking them for a list.  If they are willing to provide that, then we could include it in the nginx config or in your own firewall.

**Reply by baylanger** [reply]
It was more open discussion on how to have a bit of security on an open port. I wasn't planning to add such a list of IPs , it could change anytime w/ zero announcement from T.

For people using a service that allows to whitelist per ASN (ex. Cloudflare), that approach is probably the easiest one. Just need to confirm the "calling" IPs are part of their ASN.

IF T Fleet API always call the proxy using the path /callback , then some filtering can be done at L7. I could be wrong but per the nginx config seems that anything not /callback goes to a dead end.. so L7 filtering at router would just prevent to hit the proxy.

**Reply by baylanger** [reply]
Filtering using Tesla ASN 394161 isn't good enough. After checking logs, 1 connection came from an IP from Tesla ASN but another IP is assigned to an ASN that belongs to Amazon.

Personally, in case there's ever an odd security issue that hackers could easily find the tesla endpoint, instead of using ~ tesla.domain.com , I prefer to use a "random name".

**Reply by llamafilm** [reply]
Sure, you can use any name you want. But what risk are you concerned about, supposing a hacker finds your endpoint? In my opinion, security through obscurity isn’t worth much. If there’s any vulnerability then it should be fixed.

**Reply by baylanger** [reply]
As long as hackers don't have the exact fqdn to access the service, they can't do much. Only Tesla & Cloudflare' DNS has the fqdn info. Without the proper fqdn, L7 filtering directly at Cloudflare prevents any traffic to reach my server. I agree it's obscure security but it's good protection against DoS and if there's ever a real vulnerability, I still feel safer. Cloudflare reaches my hass server over a VPN, port 443 is fully blocked on my router.

There are hackers running "service based names" DNS queries to find services that might be up and running. "service based names", I refer to queries like www, web, cloud, san, nas, ftp, samba, ads, etc and looking at DNS log queries, it does happen quite often.
